Add save_for_forward to custom function

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71569
* #71531
* #71530

Not sure if this is the right API

Differential Revision: [D33695395](https://our.internmc.facebook.com/intern/diff/D33695395)

soulitzer committed:
(WIP) Add save_for_forward to custom function

[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Add save_for_forward to custom function"


Not sure if this is the right API


[ghstack-poisoned]

soulitzer commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -813,9 +818,20 @@ static PyObject *unpack_saved_variables(
-  return unpack_saved_variables(self, [](const Variable& var) {
-    return THPVariable_Wrap(var);
-  });
+  // return unpack_saved_variables(self, [](const Variable& var) {
```
TODO: remove this

soulitzer commented on torch/autograd/function.py:
```diff
@@ -54,6 +55,65 @@ def save_for_backward(self, *tensors: torch.Tensor):
+            if tensor is None:
+                primals.append(None)
+            else:
+                primals.append(fwAD.unpack_dual(tensor).primal)
```
I think we should unpack dual here and only provide the primals in ctx.saved_tensor. All the necessary tangents for computation are already provided as arguments, and because forward AD is not run in forward, no intermediates have tangents.

soulitzer replied:
I also prefer this solution because unpacking may be confusing (there will be two copies of tangents), and sometimes if one of the inputs does not have grad, the undefined tangent will not be materialized for the unpacked dual tensor, but will be as it is passed as input to jvp.

soulitzer replied:
Without automatic unpacking, users forgetting to unpack would also run into the following error: `RuntimeError: Setting a forward grad that itself has a forward gradient at the same level0 is not supported.`

albanD replied:
Yes unpacking sounds good!

soulitzer committed:
Update on "Add save_for_forward to custom function"


Not sure if this is the right API


[ghstack-poisoned]

soulitzer commented:
Follow up: update existing tests/code/tutorial to use save_for_forward.

soulitzer committed:
Update on "Add save_for_forward to custom function"


Not sure if this is the right API


[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33695395).

soulitzer committed:
Update on "Add save_for_forward to custom function"


Not sure if this is the right API

Differential Revision: [D33695395](https://our.internmc.facebook.com/intern/diff/D33695395)

[ghstack-poisoned]

soulitzer committed:
Update on "Add save_for_forward to custom function"


Not sure if this is the right API

Differential Revision: [D33695395](https://our.internmc.facebook.com/intern/diff/D33695395)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33695395).

soulitzer committed:
Update on "Add save_for_forward to custom function"


Not sure if this is the right API

Differential Revision: [D33695395](https://our.internmc.facebook.com/intern/diff/D33695395)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33695395).

albanD commented on torch/csrc/autograd/python_function.h:
```diff
@@ -94,6 +94,11 @@ struct THPFunction {
+    // In jvp the 'saved_tensor' attribute return tensors saved with saved_for_forward
+    // In vjp (or backward) the 'saved_tensor' attribute unpacks and returns tensors
+    // saved with saved_for_backward
+    bool is_saved_tensor_for_backward = false;
```
I am still not convinced that this flag is 100% necessary. And if it is I am not convinced by the name.

soulitzer replied:
Removed!

soulitzer commented on torch/autograd/function.py:
```diff
@@ -54,6 +55,65 @@ def save_for_backward(self, *tensors: torch.Tensor):
+
+        """
+        primals = []
+        for tensor in tensors:
```
Yep,
```
def func(*x):
    print(type(x))  # <class 'tuple'>
    print(len(x))   # 0

func()
```

soulitzer commented on torch/autograd/function.py:
```diff
@@ -54,6 +55,65 @@ def save_for_backward(self, *tensors: torch.Tensor):
+        the passed tensors need not be an input or output.
+
+        In :func:`jvp`, saved objects can be accessed through the :attr:`saved_tensors`
+        attribute. Before returning them to the user, the primals of the dual tensors
```
Yeah, if we print the dual tensor inside the forward function the tangent is not printed due to the guard....but if the user assigns the tensor to a global, they'd see that it is a dual tensor. 
So it is technically still a dual tensor no? except due to the guard, any queries to `fw_grad` return an undefined tensor.

albanD replied:
> So it is technically still a dual tensor no? except due to the guard, any queries to fw_grad return an undefined tensor.

For all intense and purposes, within that function, it only contains the dual. The user doesn't know that they are the same objects.

soulitzer committed:
Update on "Add save_for_forward to custom function"


Not sure if this is the right API

Differential Revision: [D33695395](https://our.internmc.facebook.com/intern/diff/D33695395)

[ghstack-poisoned]

soulitzer commented on torch/autograd/function.py:
```diff
@@ -54,6 +55,65 @@ def save_for_backward(self, *tensors: torch.Tensor):
+
+        This should be only called from inside the :func:`forward` method.
+        This method should only be called with tensors. Unlike `save_for_backward`,
+        the passed tensors need not be an input or output.
```
Interesting, I also updated the save_for_backward description. I know we never enforced it, but is the reason that we had the note originally (before the saved variable updates) because there was overhead from to making a shallow copy?

albanD replied:
The core idea for this is to avoid memory leak. And so this was not needed to avoid this one.
Indeed it is unnecessary overhead when you pass in non-input/output Tensors.

albanD commented on torch/autograd/function.py:
```diff
@@ -7,16 +7,16 @@
-        :func:`forward` **method. This should only be called with input or
-        output tensors**
+        This should be called at most once, and only from inside the
+        :func:`forward` method. This method should only be called with tensors.
```
Not sure we want to modify this. We don't want the user to use it for non-input/output. And it does say "should" hinting at the fact that it is not checked :P

soulitzer committed:
Update on "Add save_for_forward to custom function"


Not sure if this is the right API

Differential Revision: [D33695395](https://our.internmc.facebook.com/intern/diff/D33695395)

[ghstack-poisoned]

albanD commented on torch/autograd/function.py:
```diff
@@ -7,16 +7,21 @@
+
+        All tensors intended to be used in the backward pass should be saved
+        with ``save_for_backward`` (as opposed to directly on ``ctx``) to prevent
+        incorrect gradients and memory leaks, and enable the application of saved
```
nit: remove one "and"

soulitzer replied:
Hmm I was thinking prevent applies to incorrect gradients and memory leaks only, but perhaps I can reword this better?

albanD replied:
Maybe On the one hand to prevent incorrect gradients and memory leak, on the other hand etc  ?

soulitzer replied:
Might be too verbose actually, yeah its a bit tricky to parse in its current state but I don't see an option that is uniformly better - will just keep for now.

albanD commented on torch/autograd/function.py:
```diff
@@ -54,6 +58,57 @@ def save_for_backward(self, *tensors: torch.Tensor):
+            >>>
+            >>>     @staticmethod
+            >>>     def jvp(ctx, x_t, y_t, _):
+            >>>         x_p, y_p = ctx.saved_tensors
```
should be x, y no?

soulitzer replied:
ah good catch!

soulitzer commented:
Updated our docs to say (something along the lines of): one /should/ always store all tensors intended for use in backward with `save_for_backward`, and analogously use `save_for_forward` for forward.
Its much less confusing than to discussing case-by-case for inputs/outputs, non-inputs/non-outputs, jvp, vjp.

albanD reviewed (approved):
Looks good! Thanks for the update

soulitzer committed:
Update on "Add save_for_forward to custom function"


Not sure if this is the right API

Differential Revision: [D33695395](https://our.internmc.facebook.com/intern/diff/D33695395)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33695395).

soulitzer committed:
Update on "Add save_for_forward to custom function"


Not sure if this is the right API

Differential Revision: [D33695395](https://our.internmc.facebook.com/intern/diff/D33695395)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33695395).
